dagshub/auth/oauth.py

```diff
@@ -0,0 +1,56 @@
+import sys
+from typing import Optional, Dict
+import logging
+
+import pytimedinput
+import requests
+import urllib
+import uuid
+from dagshub.common import config
+
+
+CODE_INPUT_TIMEOUT = 60
+
+logger = logging.getLogger(__name__)
+
+
+def oauth_flow(
+    host: str,
+    client_id: Optional[str] = None,
+    code_input_timeout: Optional[int] = None,
+) -> Dict:
+    if client_id is None:
+        client_id = config.client_id
+    if code_input_timeout is None:
+        code_input_timeout = CODE_INPUT_TIMEOUT
+    state = uuid.uuid4()
+    host = host.strip("/")
+    dagshub_url = urllib.parse.urljoin(host, "login/oauth")
+    link_prompt = f"Go to {dagshub_url}/authorize?state={state}&client_id={client_id} and paste the code back in here."
+    code_prompt = "Code:"
+    if code_input_timeout <= 0:
+        print(link_prompt)
+        code = input(code_prompt)
```

Comment:
The redeem code is short-lived and supposed to be (but not sure if it is) a once-use only.
I will put a getpass in here though, you're right.
I think I used an input here because I didn't like how it shows you warnings in IntelliJ when you use it, but that's less relevant than security